cmd/backup/main.go

```diff
@@ -330,10 +332,28 @@ func (s *script) stopContainers() (func() error, error) {
 // saves it to disk.
 func (s *script) takeBackup() error {
 	s.file = timeutil.Strftime(&s.start, s.file)
-	if err := targz.Compress(s.c.BackupSources, s.file); err != nil {
+	backupSources := s.c.BackupSources
+	if s.c.BackupFromSnapshot {
+		backupCopiedSource := filepath.Join("/tmp", s.c.BackupSources)
+		backupSources = backupCopiedSource
+		s.logger.Infof("Created snapshot of `%s` at `%s`.", s.c.BackupSources, backupCopiedSource)
+		// copy before compressing guard against a situation where backup folder's content are still growing.
+		if err := copy.Copy(s.c.BackupSources, backupCopiedSource, copy.Options{ PreserveTimes: true }); err != nil {
+			return fmt.Errorf("takeBackup: error creating snapshot on backup folder: %w", err)
+		}
+		defer func() {
+			if err := os.RemoveAll(backupCopiedSource); err != nil {
+				s.logger.Errorf("takeBackup: error removing snapshot folder %s.", backupCopiedSource)
+				os.Exit(1)
```

Comment:
This is correct because we don't want to continue running the backup script anymore, but it also creates a different problem: everything else that is still "to unwind" (Go jargon for running things that are deferred before exiting) will never run in this case as the program immediately exits. This happens to include the hooks for sending out notifications on failed backups which is a little unfortunate (as this is a failed backup).

I think we have two options for fixing this:

### Make this a warning only

We could tolerate this error, log it as a warning and let the script continue to run. This could create a few weird edgecases but would probably be ok in most cases.

### Have `func main` handle all orchestration

This program already has a pattern for handling such cases, which is having a method return a cleanup function that is called no matter what. `func main` then schedules this function to run no matter if an error was returned or not.

If we wanted to apply this pattern here as well, we'd need to do the following:
- Change the signature of `(*script).takeBackup` to be `takeBackup() (func() error, error)`
- Instead of deferring the function here, return it (or return a noop in case the feature is not used)
- When calling the method in `func main` it would look something like this:
    ```go
	cleanup, err := s.takeBackup()
	defer func() {
		s.must(cleanup())
	}()
	if err != nil {
		return err
	}
   ```

The program already follows this pattern when stopping containers (which returns a function to bring them back up), so you can also check how this looks like there.

---

I would probably prefer using option 2 as it's the most correct, but it's also making this more complicated. If you feel like this is too much, let me know. Also sorry for not thinking about this in depth with my initial comment.

Reply:
is it possible if I simply move my cleanup function to `removeArtifacts` function?
```
func (s *script) removeArtifacts() error {
	_, err := os.Stat(s.file)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return fmt.Errorf("removeArtifacts: error calling stat on file %s: %w", s.file, err)
	}
	if err := os.Remove(s.file); err != nil {
		return fmt.Errorf("removeArtifacts: error removing file %s: %w", s.file, err)
	}
	s.logger.Infof("Removed local artifacts %s.", s.file)
  if s.c.BackupFromSnapshot {
    if err := os.RemoveAll(backupSources); err != nil {
      return fmt.Errorf("takeBackup: error removing snapshot folder %s.", backupSources)
    }
    s.logger.Infof("Removed snapshot folder %s.", backupSources)
  }

	return nil
}
```

Reply:
thx for the lengthy explanation, this is helpful for me to understand this project better

Reply:
> is it possible if I simply move my cleanup function to removeArtifacts function?

That's a good middle ground, I like it. In the long run such clean up could be moved into an "all-levels" exit hook, but I'd have to think about this further. For now, this solution is just right :+1:

Reply:
I did some refactoring on this topic in here https://github.com/offen/docker-volume-backup/pull/33 in case you're interested.